File: workers/typescript/apps/registry.ts
import type { App } from '../harness';
import { run } from '../harness';
import { app as workerApp } from './worker/app';

const defaultAppName = 'worker';

const registry: Record<string, App> = {
  worker: workerApp,
};

function main(argv: readonly string[] = process.argv.slice(2)): Promise<void> {
  let appName = defaultAppName;
  let index = 0;

  if (argv[0] === '--app') {
    appName = argv[1] ?? '';
    index = 2;
  } else if (argv[0]?.startsWith('--app=')) {
    appName = argv[0].slice('--app='.length);
    index = 1;
  }

  const app = registry[appName];
  if (app === undefined || appName === '') {
    throw new Error(`unknown TypeScript worker app ${JSON.stringify(appName)}`);
  }

  return run(app, argv.slice(index));
}

if (require.main === module) {
  void main()
    .then(() => {
      process.exit(0);
    })
    .catch((err) => {
      console.error(err);
      process.exit(1);
    });
}

File: workers/dotnet/Apps/Registry.cs
using Temporalio.Omes.Apps.HelloWorld;
using Temporalio.Omes.Apps.Worker;
using HarnessApp = Temporalio.Omes.Projects.Harness.App;

namespace Temporalio.Omes.Apps;

public static class Registry
{
    private const string DefaultAppName = "worker";

    private static readonly IReadOnlyDictionary<string, HarnessApp> Apps = new Dictionary<string, HarnessApp>
    {
        ["helloworld"] = HelloWorldApp.Create(),
        ["worker"] = WorkerApp.Create(),
    };

    public static Task<int> RunAsync(string[] args)
    {
        var appName = DefaultAppName;
        var harnessArgs = args;
        if (args.Length >= 2 && args[0] == "--app")
        {
            appName = args[1];
            harnessArgs = args[2..];
        }

        if (appName.Length == 0 || !Apps.TryGetValue(appName, out var app))
        {
            Console.Error.WriteLine($"unknown .NET worker app {appName}");
            return Task.FromResult(1);
        }

        return HarnessApp.RunAsync(app, harnessArgs);
    }
}
